### PR TITLE
[Merge/release] Accommodate custom .CUSTOMERS via ENV var

### LINKED ISSUE #52: Allow custom URL via env-var for customer validation
## Expected Behaviour

Anyone can host their own Derek and also validate their own customer list

## Current Behaviour

The code will validate against [our list](https://github.com/alexellis/derek/blob/d1dc963fea622554ae940f0892427cc6cde54dea/auth/customers.go#L21).

## Possible Solution

There is a workaround though - just disable customer validation via the existing env-var flag. 

I'd suggest we allow the customer list URL to be supplied via env-var too.

Related #51

### PR DESCRIPTION
**Pull request by rgee0**
The existing location of the .CUSTOMERS file is hard-coded which means
that anyone self hosting will need to either add an entry into the project
.CUSTOMERS, or alter the source code to point at their own location.

This change enables an over-ride of the default .CUSTOMERS location by
setting an `customers_url` environment variable.  The location should
be an `https` endpoint.

Test added for the new function which attempts to read the ENV var and
GET.md updated to include `customers_url` in the stack file section.

Signed-off-by: rgee0 <richard@technologee.co.uk>

## Motivation and Context
- [x] I have raised an issue to propose this change (fixes #52)

## How Has This Been Tested?

Tested a number of scenarios on a DigitalOcean machine.

* Ran in default mode and confirmed as functioning as intended
* Ran with `customers_url: https://raw.githubusercontent.com/rgee0/Test/master/.CUSTOMERS` which **didn't** contain `rgee0`: 

```
derek.1.zcextiz4jcr8@derek    | 2018/04/22 19:07:40 Forking fprocess.
derek.1.zcextiz4jcr8@derek    | 2018/04/22 19:07:40 Query  
derek.1.zcextiz4jcr8@derek    | 2018/04/22 19:07:40 Path  /function/derek
derek.1.zcextiz4jcr8@derek    | 2018/04/22 19:07:40 stderr: time="2018-04-22T19:07:40Z" level=fatal msg="No customer found for: rgee0/Test"
derek.1.zcextiz4jcr8@derek    | 2018/04/22 19:07:40 Success=false, Error=exit status 1
derek.1.zcextiz4jcr8@derek    | 2018/04/22 19:07:40 Out=
```
* Ran with `customers_url: raw.githubusercontent.com/rgee0/Test/master/.CUSTOMERS` which **didn't** contain `rgee0`: 

```
derek.1.l41j9x9d4tnz@derek    | 2018/04/22 19:11:19 Forking fprocess.
derek.1.l41j9x9d4tnz@derek    | 2018/04/22 19:11:19 Query  
derek.1.l41j9x9d4tnz@derek    | 2018/04/22 19:11:19 Path  /function/derek
derek.1.l41j9x9d4tnz@derek    | 2018/04/22 19:11:19 stderr: time="2018-04-22T19:11:19Z" level=fatal msg="No customer found for: rgee0/Test"
derek.1.l41j9x9d4tnz@derek    | 2018/04/22 19:11:19 Success=false, Error=exit status 1
derek.1.l41j9x9d4tnz@derek    | 2018/04/22 19:11:19 Out=
```
* Ran with `customers_url: raw.githubusercontent.com/rgee0/Test/master/.CUSTOMERS` which **did** contain `rgee0`: 

```
derek.1.l41j9x9d4tnz@derek    | 2018/04/22 19:12:55 Forking fprocess.
derek.1.l41j9x9d4tnz@derek    | 2018/04/22 19:12:55 Query  
derek.1.l41j9x9d4tnz@derek    | 2018/04/22 19:12:55 Path  /function/derek
derek.1.l41j9x9d4tnz@derek    | rgee0 wants to add label of 'fred' on issue #8 
derek.1.l41j9x9d4tnz@derek    | 2018/04/22 19:12:56 Duration: 1.263511 seconds
```
* Ran with `customers_url: https://raw.githubusercontent.com/rgee0/Test/master/.CUSTOMERS` which **did** contain `rgee0`: 
```
derek.1.ty4gf05lfayb@derek    | 2018/04/22 19:14:35 Forking fprocess.
derek.1.ty4gf05lfayb@derek    | 2018/04/22 19:14:35 Query  
derek.1.ty4gf05lfayb@derek    | 2018/04/22 19:14:35 Path  /function/derek
derek.1.ty4gf05lfayb@derek    | rgee0 wants to add label of 'fred' on issue #8 
derek.1.ty4gf05lfayb@derek    | 2018/04/22 19:14:36 Duration: 1.042601 seconds
```
* Ran with `customers_url: https://raw.githubusercontent.com/rgee0/Test/master/.CONSUMERS` which **didnt** exist: 
```
derek.1.z8r0vvm6dk96@derek    | 2018/04/22 19:22:41 Forking fprocess.
derek.1.z8r0vvm6dk96@derek    | 2018/04/22 19:22:41 Query  
derek.1.z8r0vvm6dk96@derek    | 2018/04/22 19:22:41 Path  /function/derek
derek.1.z8r0vvm6dk96@derek    | 2018/04/22 19:22:41 stderr: time="2018-04-22T19:22:41Z" level=fatal msg="No customer found for: rgee0/Test"
derek.1.z8r0vvm6dk96@derek    | 2018/04/22 19:22:41 Success=false, Error=exit status 1
derek.1.z8r0vvm6dk96@derek    | 2018/04/22 19:22:41 Out=
```

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I've read the [CONTRIBUTION](https://github.com/alexellis/derek/blob/master/CONTRIBUTING.md) guide
- [x] I have signed-off my commits with `git commit -s`
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by alexellis** [comment]
Thank you for working on this Richard. It looks like you have most of this ready to go but I have a bit of feedback on the naming. 

> This change enables an over-ride of the default .CUSTOMERS location by
> setting an `customers_yaml` environment variable. The location should
> be a https endpoint.

I wonder if by setting the name to `customers_url` we could avoid any ambiguity? The destination file isn't YAML at the moment either - it's entries separated by new-line as I remember.

(I also saw some references to customerYAML etc in the code.)

Please could you make some wording changes and push again when you next have time to re-test?

File: auth/customers_test.go [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+			expectedURL: defaultCustomersURL,
+		},
+		{
+			title:       "ENV var exists",
+			envVar:      "https://raw.githubusercontent.com/rgee0/derek/master/.CUSTOMERS",
+			expectedURL: "https://raw.githubusercontent.com/rgee0/derek/master/.CUSTOMERS",
+		},
+		{
+			title:       "Invalid ENV var exists",
+			envVar:      "raw.githubusercontent.com/rgee0/derek/master/.CUSTOMERS",
```
[the remark this answers is not in the document]
Comment (rgee0) [reply auth/customers_test.go]: Stranger things have happened. HTTPS will be forced in that event and the README is clear in stating a HTTPS endpoint is expected.

File: auth/customers.go [hunk 2]
```diff
@@ -18,7 +35,8 @@ func IsCustomer(repo types.Repository) (bool, error) {
 	var err error
 	var found bool
 	c := http.Client{}
-	request, _ := http.NewRequest(http.MethodGet, "https://raw.githubusercontent.com/alexellis/derek/master/.CUSTOMERS", nil)
+	customersURL := findCustomersURL()
+	request, _ := http.NewRequest(http.MethodGet, customersURL, nil)
 	res, doErr := c.Do(request)
 	if err != nil {
```
Comment (rgee0) [inline auth/customers.go]: @alexellis is this an error?  I was expecting `doErr` to be evaluated in the `if`?

Comment (alexellis) [reply auth/customers.go]: Yes that's a bug waiting to happen. Can you fix?

File: auth/customers_test.go [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+			expectedURL: "https://raw.githubusercontent.com/rgee0/derek/master/.CUSTOMERS",
+		},
+	}
+
+	for _, test := range URIOpts {
```
Comment (alexellis) [inline auth/customers_test.go]: I won't be picky but I think go likes these things to be written like `uriOpts`

File: auth/customers.go [hunk 4]
```diff
@@ -9,6 +10,22 @@ import (
...
 )
 
+const defaultCustomersURL string = "https://raw.githubusercontent.com/alexellis/derek/master/.CUSTOMERS"
+const customersURLEnv string = "customers_url"
+
+func findCustomersURL() string {
+
+	if customURL, exists := os.LookupEnv(customersURLEnv); exists && (len(customURL) > 0) {
+
+		if !strings.HasPrefix(strings.ToLower(customURL), "http") {
```
Comment (alexellis) [inline auth/customers.go]: I don't know why we need this prepending of https?

If the customers URL doesn't have a prefix of http/https we should probably throw an error.

File: auth/customers_test.go (hunk 3 above)
Comment (rgee0) [reply auth/customers_test.go]: https://github.com/golang/go/wiki/CodeReviewComments#initialisms

Comment (alexellis) [reply auth/customers_test.go]: From linked page:

> For example, "URL" should appear as "URL" or "url" (as in "urlPony", or "URLPony"

The linked page implies both are valid. The current codebase uses the first variant. Doesn't need to change

File: auth/customers.go (hunk 4 above)
Comment (alexellis) [reply auth/customers.go]: This I think can be removed though. I'm not sure what the usecase is for it.

**Pull request merged**